new: update to handle network connection objects

emmekappa86 opened the pull request:
#### What does it do?

This update concerns the possibility of creating snort/suricata rules from network-connection objects using multiple attributes (instead of splitting those attributes into a series of unrelated rules).
This change was mentioned during the CONCORDIA MISP Training (May 27th 2021) and authored by (me) Marco Caselli from Siemens AG.

#### Questions

- [ ] Does it require a DB change?
- [ ] Are you using it in production?
- [ ] Does it require a change in the API (PyMISP for example)?

emmekappa86 committed:
new: update to handle network connection objects

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -16,11 +16,14 @@ class NidsExport
 			)
 		),
-		'flatten' => 1
+		#'flatten' => 1
```
Any debugging / test commented lines should be removed

emmekappa86 replied:
Done.

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -16,11 +16,14 @@ class NidsExport
 	public function handler($data, $options = array())
 	{
+	
+		//NOTES: Here the scope "Object" should be probably checked
```
Not sure I understand the meaning of this comment. However, it should be noted that the `network-connection` object is only supported when the scope `event` is used. Would it be worth investing more time trying to have the same feature while using the `attribute` scope?

emmekappa86 replied:
The comment should have been deleted as it was written at the very beginning while studying the code. However, I am not sure I get exactly the situation in which an "object" signature will be created while using the "attribute" scope. Could you make an example?

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -38,15 +41,16 @@ public function handler($data, $options = array())
-				foreach ($data['Object'] as $object) {
-					$this->__convertFromEventFormat($object['Attribute'], $data, $options, $continue);
-				}
+				#foreach ($data['Object'] as $object) {
```
Delete line?

emmekappa86 replied:
Done.

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -69,6 +73,44 @@ private function __convertFromEventFormat($attributes, $event, $options = array(
+	
+	private function __convertFromEventFormatObject($objects, $event, $options = array(), $continue = false) {
+		
+		#CakeLog::debug("ConvertFromEventFormatObject");
```
Delete line?

emmekappa86 replied:
Done.

mokaddem commented:
Hello!
Thanks a lot for the improvements.
In addition to the few comments placed in the PR itself, please find some general comments below:
1. The biggest issue I see is that when using the `event` scope (`/events/restSearch`), any `attributes` belonging to an object will be ignored if the object is not of the type `network-connection`. This is an issue as we have lots of different objects (`domain-ip`, ...) that should also be included in the export. This is not an issue for the `attribute` `scope` (see point 2).
2. Not sure if it's done on purpose or not but `network-connection` objects are only handled as intended when the `scope` is `event`. Not while using `attribute` (See related comment inline of this PR).
3. Also here is a mix between tabs and spaces in the file. Could it be possible to uniform it? (Preferably with spaces?).

emmekappa86 commented:
Hi mokaddem,

Thanks for the feedback and sorry for my late response.
I will go through your comments and suggestions and try to fix everything.

Best regards

             Marco

emmekappa86 committed:
Merge branch 'MISP:2.4' into feature-snort-rule-from-network-connection

emmekappa86 committed:
fixes + ddos object handling

emmekappa86 committed:
Code polishing

emmekappa86 commented:
Hi mokaddem,

Here is my update:

1.  I have changed the code to ensure that any object not included in the "supportedObject" list is treated as it used to be (thus, considering the attributes one by one
2. As written in the comments above, I might not have fully understood the way scopes are used. What I thought was that the feature was needed only when the scope is "Event". Therefore, I have implemented it this way. Furthermore, I tried to follow the previous code (the objects were only checked within the "Event" scope)
3. I replaced all tabs with spaces

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -7,84 +7,125 @@ class NidsExport
+                );
+
+            } else { // In case no custom export exists for the object, the approach falls back to the attribute case
+                $this->__convertFromEventFormat($object['Attribute'], $data, $options, $continue);
```
Replace `$data` by `$event`

emmekappa86 replied:
Done.

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -142,6 +183,9 @@ public function export($items, $startSid, $format="suricata", $continue = false)
             $sid = $startSid + ($item['Attribute']['id'] * 10); // leave 9 possible rules per attribute type
             $sid++;
+            
+            if(!empty($item['Attribute']['type'])) { // item is an 'Attribute'
```
Fix indentation

emmekappa86 replied:
Done.

mokaddem commented on app/Lib/Export/NidsExport.php:
```diff
@@ -201,10 +245,116 @@ public function export($items, $startSid, $format="suricata", $continue = false)
                     break;
             }
+
+          } else if(!empty($item['Attribute']['name'])) { // Item is an 'Object'
```
Fix indentation

emmekappa86 replied:
Done.

mokaddem commented:
Hello @emmekappa86,
Add a look at your changes and I still have some nitpicking.
The rest looks good.
Looking forward for these fixes to merge the PR!

emmekappa86 committed:
Fixing mistake ("data" -> "event")

emmekappa86 committed:
Fixed indentation

emmekappa86 commented:
Thanks for the review. I fixed the reported mistakes.

emmekappa86 commented:
Hi @mokaddem, I wanted to ask if this feature can be merged. Is there anything else I should do/fix in the code?

mokaddem commented:
Hey, finally merged it into the [`develop` branch](https://github.com/MISP/MISP/commit/8f4c850cde2b5296b05efa2ded21f0e4aee7466a)
Thanks a lot for the work!
